cypress/support/PersistentLogs.js

```diff
@@ -16,15 +21,98 @@ console.log = function (message) {
+
+// FINAL LOGGING SUPPORT FOR PREVIOUS TEST CASE
+// In an ideal world this code would be in an afterEach call, however, since 
+// there will be other afterEach functions that will be called after ours and 
+// we want to ensure the logging we do here is the final logging for the test 
+// case.
+//
+// Also important to notice is that this results in a call to cy.writeFile
+// which will execute in a different time sequence than all of the other 
+// JavaScript in this code block. Thus in this code block we cannot run any
+// true `beforeEach` code that is related to the current test case that is
+// just about to start.
+// FINAL LOGGING SUPPORT FOR PREVIOUS TEST CASE
+beforeEach(function() {
+  LogTestState()
+  FlushLogEntries()
+})
+
+// TRUE `beforeEach` CODE FOR THE TEST CASE THAT IS JUST NOW STARTING
+// This MUST come after the prior `beforeEach` code block and cannot
+// be combined with that other code block due to the `cy.writeFile`
+// call that it contains.
+beforeEach(function() {
+  console.log('')
+  console.log(`******************** ${GetFullTestSuiteTitle(this.currentTest)} ************************************************************`)
+
+  test = this.currentTest
+  let title = GetSuiteRootTitle(this.currentTest)
+  if (title !== suiteTitle) {
+    suiteTitle = title
+    logFileName = `./results/cypress/${suiteTitle}.${Cypress.moment().format("YY.MM.DD.HH.mm.ss..SSS")}.log`
+  }
+})
 
 // After each test case ends...
-afterEach(() => {
-  let logFileName = `./results/cypress/${testCaseRegistry[indexRegistry++]}.${Cypress.moment().format("YY.MM.DD.HH.mm.ss..SSS")}.log`
-  cy.writeFile(logFileName, logEntries).then(() => logEntries = '')
+afterEach(() => { WriteLogEntries() })
+
+after(() => { 
+  LogTestState()
+  FlushLogEntries() 
 })
+
+// -----------------------------------------------------------------------------
+// Support functions.
+// -----------------------------------------------------------------------------
+
+function LogTestState()
+{
+  if (!test) { return false }
+
+  let message = `: ${test.state.toUpperCase()} - Test Case: '${GetFullTestSuiteTitle(test)}' - ${test.state.toUpperCase()} :`
+  console.log('.'.repeat(message.length))
+  console.log(message)
+  console.log("'".repeat(message.length))
+}
+
+// NOTE: This requires the latest version (3.1.5) of Cypress:
+//        change directory to the root project folder
+//        npm install cypress --save-dev
+function WriteLogEntries() { 
+  if (logFileName) {
+    cy.writeFile(logFileName, logEntries, {flag: 'a'})
+      .then(() => { logEntries = '' })
+  }
+} 
+
+function FlushLogEntries() {   
+  if (logFileName && logEntries.length > 0) {
+    // Because of 'afterEach' functions that were called after we wrote our previous
+    // log entries, there are additional log entries to persist to that same log file.
+    WriteLogEntries()
+  }
+}
+
+function GetSuiteRootTitle(test) {
+  if (!test.parent) throw new 'Did not find test.parent'
```

Comment:
Does the code throw new 'Did not find test.parent' actually run without causing it's own error?

I didn't think you could follow the new keyword with a string. new is for creating new objects from constructor functions like: const myInstance = new MyClass() . I don't know what new "asbasdf" does.

In any case, always throw actual Error objects to ensure you get the proper stack trace and the consumers of errors are guaranteed the format. Should also make the error messages as helpful/informative as possible.

E.g. throw new Error(`There is no parent for test: ${test.title}`) tells them which test didn't have a parent to help investigating.

Reply:
The new is a mistake...a throwback to my C# days.

Good point on adding `test.title` to the error, I'll do it.